# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

AdjustAlleleFreq <- function(predAl, alleles2loc, minfreq) {
    .Call('_polyRAD_AdjustAlleleFreq', PACKAGE = 'polyRAD', predAl, alleles2loc, minfreq)
}

BestGenos <- function(probs, ploidy, ntaxa, nalleles) {
    .Call('_polyRAD_BestGenos', PACKAGE = 'polyRAD', probs, ploidy, ntaxa, nalleles)
}

CorrectGenos <- function(bestgenos, probs, alleles2loc, ntaxa, ploidy, nalleles, nloc, do_correct) {
    .Call('_polyRAD_CorrectGenos', PACKAGE = 'polyRAD', bestgenos, probs, alleles2loc, ntaxa, ploidy, nalleles, nloc, do_correct)
}

BestPloidies <- function(chisq) {
    .Call('_polyRAD_BestPloidies', PACKAGE = 'polyRAD', chisq)
}

FormatStructure <- function(genotypes, alleles2loc, ploidy) {
    .Call('_polyRAD_FormatStructure', PACKAGE = 'polyRAD', genotypes, alleles2loc, ploidy)
}

GiniSimpson <- function(counts) {
    .Call('_polyRAD_GiniSimpson', PACKAGE = 'polyRAD', counts)
}

HindHeMat <- function(alleleDepth, depthRatio, alleles2loc, nLoci, He) {
    .Call('_polyRAD_HindHeMat', PACKAGE = 'polyRAD', alleleDepth, depthRatio, alleles2loc, nLoci, He)
}

HoOneParent <- function(genotypes, alleles2loc, keeploc, ploidy) {
    .Call('_polyRAD_HoOneParent', PACKAGE = 'polyRAD', genotypes, alleles2loc, keeploc, ploidy)
}

HoTwoParents <- function(genotypes1, genotypes2, alleles2loc, keeploc, ploidy1, ploidy2) {
    .Call('_polyRAD_HoTwoParents', PACKAGE = 'polyRAD', genotypes1, genotypes2, alleles2loc, keeploc, ploidy1, ploidy2)
}

InitHapAssign <- function(NMmat) {
    .Call('_polyRAD_InitHapAssign', PACKAGE = 'polyRAD', NMmat)
}

Hap2SNP <- function(haps, refhap, pos) {
    .Call('_polyRAD_Hap2SNP', PACKAGE = 'polyRAD', haps, refhap, pos)
}

Hap2Hap <- function(haps, refhap, pos) {
    .Call('_polyRAD_Hap2Hap', PACKAGE = 'polyRAD', haps, refhap, pos)
}

MakeGTstring <- function(genotype, ploidy) {
    .Call('_polyRAD_MakeGTstring', PACKAGE = 'polyRAD', genotype, ploidy)
}

PrepVCFexport <- function(genotypes, alleles2loc, alleleDepth, alleleNucleotides, locTable, ploidy, taxaPloidy, asSNPs) {
    .Call('_polyRAD_PrepVCFexport', PACKAGE = 'polyRAD', genotypes, alleles2loc, alleleDepth, alleleNucleotides, locTable, ploidy, taxaPloidy, asSNPs)
}

simGeno <- function(alleleFreq, alleles2loc, nsam, inbreeding, ploidy) {
    .Call('_polyRAD_simGeno', PACKAGE = 'polyRAD', alleleFreq, alleles2loc, nsam, inbreeding, ploidy)
}

simGenoMapping <- function(donorGeno, recurGeno, progGeno, genoProbs, alleles2loc, nsam, ploidyDon, ploidyRec) {
    .Call('_polyRAD_simGenoMapping', PACKAGE = 'polyRAD', donorGeno, recurGeno, progGeno, genoProbs, alleles2loc, nsam, ploidyDon, ploidyRec)
}

simAD <- function(locDepth, genotypes, alleles2loc, overdispersion, contamRate, alleleFreq, errorRate) {
    .Call('_polyRAD_simAD', PACKAGE = 'polyRAD', locDepth, genotypes, alleles2loc, overdispersion, contamRate, alleleFreq, errorRate)
}

ThirdDimProd <- function(probs, ngen, ntaxa) {
    .Call('_polyRAD_ThirdDimProd', PACKAGE = 'polyRAD', probs, ngen, ntaxa)
}

